### ISSUE TITLE
How to document a project with files in source hierarchy with mdbook

### ISSUE THREAD

**Issue by andrewdavidmackenzie** [rust-lang/mdBook#1024]
I have a project with a number of markdown files throughout the source code hierarchy, and I also have a "book" in a docs directory.

```
+---subproject1
|    +---README.md
+---subproject2
|    +---README.md
+---docs
     +--book.toml
     +--src/
        +--SUMMARY.md
```

I would like to document this project using mdbook, with the ability to refer to markdown files in the source directories from markdown files under /docs.

I understand (and tests seem to confirm it) that I cannot refer to markdown files via links that are NOT under the "src" directory specified for the book. This doesn't produce a build failure, but the links fail in the deployed book on gh-pages.

My Hack
=
So, I have a Makefile to build it that copies the markdown files *outside* the docs directory into the docs/src directory, preserving directory structure, and then I can refer to them using links in the other markdown files I write in my book.

Resulting in this:
```
+---subproject1
|    +---README.md
+---subproject2
|    +---README.md
+---docs
     +--book.toml
     +--src/
        +--SUMMARY.md
        +---subproject1
         |    +---README.md
        +---subproject2
        |    +---README.md
```

Alternative
=
So, I tried to organize my project where the "src" file is the root, and I should be able to refer to any file in the entire hierarchy.

```
+---subproject1
|    +---README.md
+---subproject2
|    +---README.md
+--book.toml
      (src = ".")
+--SUMMARY.md
```

In the docs on [mdbook](https://rust-lang-nursery.github.io/mdBook/cli/build.html) it states: 
`Note: Make sure to run the build command in the root directory and not in the source directory.`

When I try to build this, it fills the disk on my machine (:-) ), which is probably why it's recommended against.

Is there a way to do this without my original hack?

**Comment by ehuss** [comment]
You can refer to files outside of the book, it just depends on how it is deployed.  For example, the rust-lang books all deploy to https://doc.rust-lang.org/.  Each book can do relative links to other things outside of the book with relative links. For example, [the reference](https://doc.rust-lang.org/reference/) links to the [standard library docs](https://doc.rust-lang.org/std/) with links like `../std/primitive.char.html`.  The exact structure will depend on how you deploy your documentation.

So, I think for your example, you would keep your original structure, and have a build script which builds the book, and then copies all the relevant pieces to a directory for final packaging and deployment.

Alternatively you can `{{#include}}` files from outside the book as well.

**Comment by andrewdavidmackenzie** [comment]
OK, thanks for reply. 

I was trying to link to an external md file in the hope it would be built also (same FS, same "project" even if outside of the book directory), but that doesn't work.

I notice your examples links to an html file.
How is that other file built? 

In my case the `mdbook build` wouldn't built the html from the md (right?) and I don't have another book build to build it. 

I sas the include option and will use it here, but I wish to link to another md file, and once build have that link be to the built html file - just like all the files inside the "src" directory.

**Comment by ehuss** [comment]
The other files can be built using any tool. In the `std` case it is `rustdoc`. Or it can be other books built with mdbook. Or it can be buit with jekyll if it is a github site.  Or it can be hand written html.

I think for now, `{{include}}` might be your best bet.  I doubt mdbook will ever try to build pages linked to from outside of its source directory.

**Comment by andrewdavidmackenzie** [comment]
OK, that kind of makes sense....where would you stop.

What about the simple option, of the root directory having book.toml and
SUMMARY.md and being configurable as the “src” directory and mdbook being
able to build it all without some infinite loop that fills the disk?

**Comment by andrewdavidmackenzie** [comment]
OK, I see this is due to the call to `copy_files_except_ext()` in the html renderer, causing a recursive copy that fills the file system.

Would you accept a PR that excluded the build directory itself (the destination of the copy) from such a copy, avoiding recursion - and opening up the ability to "anchor" a book in the root directory of a project and solving my issue (unless there is a command option I haven't see to have the same effect?)?

**Comment by andrewdavidmackenzie** [comment]
So, I have a PR ready that would change the behaviour, to avoid a never-ending recursive copy where "build_dir" is under "src" directory.

I think it "doesn't do any harm" as such, even if the current recommendation is to *not* structure your project such. And it effectively avoids the problem warned against in the docs `Note: Make sure to run the build command in the root directory and not in the source directory.` which if not spotted causes people to fill their disk with a recursive copy.

WDYT?

If acceptable, I could consider submit enhancements to Book ("book.toml"), like an table of directories under `root_dir` or `src` to *exclude* from the build... allowing controlling what gets copied during the build and avoid copying a lot of stuff that is not wanted in the build.

**Comment by andrewdavidmackenzie** [comment]
For the ignore functionality I would use the existing "ignore" crate from the ripgrep project, and using the provided method to override the name of the ignore file from ".gitignore" to ".mdbookignore"

**Comment by andrewdavidmackenzie** [comment]
Fixed mostly by #1026 

I will submit another more focused issue on the topic of ignoring undesired files in the copy action.
